### PR TITLE
Add parser implementation

### PR DESCRIPTION
**Pull request by abenso**
Added spawn parser for Wallet, Multisig, Vesting, and Vault accounts.
Added spend parser.
Implemented cpp and zemu tests.

### CODE REVIEW

File: app/src/parser.c [hunk 1]
```diff
@@ -43,31 +43,44 @@ parser_error_t parser_init_context(parser_context_t *ctx, const uint8_t *buffer,
...
 }
 
 parser_error_t parser_getNumItems(const parser_context_t *ctx, uint8_t *num_items) {
-    // #{TODO} --> function to retrieve num Items
-    // *num_items = _getNumItems();
-    UNUSED(ctx);
-    *num_items = 2;
+    *num_items = 0;
+
+    switch (ctx->tx_obj->account_type) {
```
Comment (jleni) [inline app/src/parser.c:69]: mmm.. any chance tx_obj is null? maybe we can check?

File: app/src/parser.c [hunk 2]
```diff
@@ -93,27 +106,241 @@ parser_error_t parser_getItem(const parser_context_t *ctx, uint8_t displayIdx, c
...
 
+parser_error_t printMultisigSpawn(const parser_context_t *ctx, uint8_t displayIdx, char *outKey, uint16_t outKeyLen,
+                                  char *outVal, uint16_t outValLen, uint8_t pageIdx, uint8_t *pageCount) {
+    char buff[50] = {0};
     switch (displayIdx) {
         case 0:
-            snprintf(outKey, outKeyLen, "Blind");
-            snprintf(outVal, outValLen, "Signing");
+            snprintf(outKey, outKeyLen, "Tx type");
+            snprintf(outVal, outKeyLen, "Spawn");
```
Comment (jleni) [inline app/src/parser.c:233]: we may need to ask if they prefer to see "Multisig Spawn" or similar

File: app/src/parser.c [hunk 3]
```diff
@@ -93,27 +106,241 @@ parser_error_t parser_getItem(const parser_context_t *ctx, uint8_t displayIdx, c
...
+    return parser_ok;
+}
+
+parser_error_t printVaultSpawn(const parser_context_t *ctx, uint8_t displayIdx, char *outKey, uint16_t outKeyLen,
+                               char *outVal, uint16_t outValLen, uint8_t pageIdx, uint8_t *pageCount) {
+    char buff[50] = {0};
+    switch (displayIdx) {
+        case 0:
+            snprintf(outKey, outKeyLen, "Tx type");
+            snprintf(outVal, outKeyLen, "Spawn");
```
Comment (jleni) [inline app/src/parser.c:281]: Vault spawn?

File: app/src/parser.c [hunk 4]
```diff
@@ -93,27 +106,241 @@ parser_error_t parser_getItem(const parser_context_t *ctx, uint8_t displayIdx, c
...
+            break;
         case 1:
-            snprintf(outKey, outKeyLen, "Txn hash");
-            pageStringHex(outVal, outValLen, (char *)txnHash, sizeof(txnHash), pageIdx, pageCount);
+            snprintf(outKey, outKeyLen, "Principal");
+            pageStringHex(outVal, outValLen, (const char *)(ctx->tx_obj->principal.ptr), ADDRESS_LENGTH, pageIdx, pageCount);
```
Comment (jleni) [inline app/src/parser.c:285]: I am worried about so many pointer dereferencing without checks.

File: app/src/parser.c [hunk 5]
```diff
@@ -93,27 +106,241 @@ parser_error_t parser_getItem(const parser_context_t *ctx, uint8_t displayIdx, c
...
+            pageString(outVal, outValLen, buff, pageIdx, pageCount);
+            break;
+        case 5:
+            snprintf(outKey, outKeyLen, "Owner");
+            pageStringHex(outVal, outValLen, (const char *)(ctx->tx_obj->spawn.vault.owner.ptr), ADDRESS_LENGTH, pageIdx,
+                          pageCount);
+            break;
+        case 6:
+            snprintf(outKey, outKeyLen, "TotalAmount");
+            if (uint64_to_str(buff, sizeof(buff), ctx->tx_obj->spawn.vault.totalAmount) != NULL) {
```
Comment (jleni) [inline app/src/parser.c:314]: number of decimals?
At least we need a TODO here.. and any other place where we are missing the conversion

File: app/src/parser_impl.c [hunk 6]
```diff
@@ -16,12 +16,27 @@
...
+#include "crypto_helper.h"
+#include "scale_helper.h"
 #include "zxerror.h"
 
 parser_error_t _read(parser_context_t *c, parser_tx_t *v) {
-    UNUSED(c);
-    UNUSED(v);
-    // #{TODO} --> parse parameters: read from c->buffer and store in v
+    // read common params
+    CHECK_ERROR(_readTxVersion(c, &v->tx_version));
```
Comment (jleni) [inline app/src/parser_impl.c:25]: a check for v? ideally both

File: app/src/parser_impl.c [hunk 7]
```diff
@@ -16,12 +16,27 @@
...
+    CHECK_ERROR(readFixedArray(c, &v->principal, ADDRESS_LENGTH));
+    CHECK_ERROR(_readMethodSelector(c, &v->methodSelector));
+    switch (v->methodSelector) {
+        case METHOD_SPAWN:
+            CHECK_ERROR(_readSpawnTx(c, v));
+            break;
+        case METHOD_SPEND:
+            CHECK_ERROR(_readSpendTx(c, v));
+            break;
+    }
```
Comment (jleni) [inline app/src/parser_impl.c:35]: default: + error?

File: app/src/parser_impl_common.c [hunk 8]
```diff
@@ -0,0 +1,183 @@
...
+            if (val->spawn.vault.initialUnlockAmount > val->spawn.vault.totalAmount) {
+                return parser_unexpected_value;
+            }
+            CHECK_ERROR(readCompactU32(ctx, &val->spawn.vault.vestingStart));
+            CHECK_ERROR(readCompactU32(ctx, &val->spawn.vault.vestingEnd));
+            if (val->spawn.vault.vestingStart > val->spawn.vault.vestingEnd) {
+                return parser_unexpected_value;
+            }
+            break;
+    }
```
Comment (jleni) [inline app/src/parser_impl_common.c:171]: deafult case and error?

**Review by jleni — CHANGES_REQUESTED** [review]

File: app/src/parser.c (hunk 2 above)
Comment (abenso) [reply app/src/parser.c:233]: yes, this can help to identify the type of account. I will add this

File: app/src/parser.c (hunk 4 above)
Comment (abenso) [reply app/src/parser.c:285]: All these pointers are stored in [app/src/parser_impl_common.c](https://github.com/Zondax/ledger-spacemesh/pull/14/files/507eb7efe0d877b91ef093b894ea28b8df969729#diff-d666ddfa9baa49b477e415987a1a1b948f5e71803d1a8a05eb49bae3ed747872). At this point, all data must be validated. I can add a pointer checker at this level if you consider it important.

File: app/src/parser.c (hunk 5 above)
Comment (abenso) [reply app/src/parser.c:314]: In space mesh base code, this values are integers but I will add a TODO

**Pull request merged**